types/xdg_shell/wlr_xdg_surface.c

```diff
@@ -524,13 +524,22 @@ void wlr_xdg_toplevel_send_close(struct wlr_xdg_surface *surface) {
 }
 
 void wlr_xdg_popup_destroy(struct wlr_xdg_surface *surface) {
+	if (surface == NULL) {
+		return;
+	}
 	assert(surface->popup);
-	assert(surface->role == WLR_XDG_SURFACE_ROLE_POPUP);
+	assert(surface->role == WLR_XDG_SURFACE_ROLE_POPUP ||
+		surface->role == WLR_XDG_SURFACE_ROLE_NONE);
 	xdg_popup_send_popup_done(surface->popup->resource);
+	wl_resource_set_user_data(surface->popup->resource, NULL);
+	destroy_xdg_popup(surface->popup->base);
 }
 
 static void xdg_popup_get_position(struct wlr_xdg_popup *popup,
 		double *popup_sx, double *popup_sy) {
+	if (popup == NULL) {
```

Comment:
NULL checks should be added in protocol request handlers, not here. That means `xdg_popup_handle_grab`, but apparently that's already done, so we're good.